File: examples/chapter02/mcp-server_qiita/python/tests/integration_test.py
"""Qiita API統合テスト

注意: このテストは実際のQiita APIを呼び出すため、以下の点に注意してください:
- レート制限（60リクエスト/時）を考慮して実行してください
- ネットワーク接続が必要です
- テスト間に待機時間を設けています
"""

import asyncio

import pytest

from src.qiita import search_qiita


@pytest.mark.asyncio
async def test_search_single_keyword() -> None:
    """単一キーワードでの検索テスト"""
    articles = await search_qiita(keywords=["TypeScript"], per_page=5, page=1)

    assert isinstance(articles, list)

    if len(articles) > 0:
        article = articles[0]

        # 必須フィールドの存在確認
        assert hasattr(article, "id")
        assert hasattr(article, "title")
        assert hasattr(article, "url")
        assert hasattr(article, "body")
        assert hasattr(article, "created_at")
        assert hasattr(article, "updated_at")
        assert hasattr(article, "likes_count")
        assert hasattr(article, "stocks_count")
        assert hasattr(article, "user")
        assert hasattr(article, "tags")

        # ユーザー情報の確認
        assert hasattr(article.user, "id")
        assert hasattr(article.user, "name")
        assert hasattr(article.user, "profile_image_url")

        # タグ情報の確認
        assert isinstance(article.tags, list)

        if len(article.tags) > 0:
            tag = article.tags[0]
            assert hasattr(tag, "name")
            assert hasattr(tag, "versions")
            assert isinstance(tag.versions, list)

    # レート制限を考慮して待機
    await asyncio.sleep(3)


@pytest.mark.asyncio
async def test_search_multiple_keywords() -> None:
    """複数キーワードでの検索テスト"""
    articles = await search_qiita(keywords=["MCP", "API"], per_page=5, page=1)

    assert isinstance(articles, list)

    # 結果が返ってきた場合、各記事の型が正しいことを確認
    for article in articles[:3]:  # 最初の3件のみチェック
        assert hasattr(article, "id")
        assert isinstance(article.id, str)
        assert hasattr(article, "title")
        assert isinstance(article.title, str)
        assert hasattr(article, "url")
        assert isinstance(article.url, str)


@pytest.mark.asyncio
async def test_search_with_pagination() -> None:
    """ページネーションのテスト"""
    # 1ページ目を取得
    articles_page1 = await search_qiita(keywords=["Python"], per_page=3, page=1)

    await asyncio.sleep(3)

    # 2ページ目を取得
    articles_page2 = await search_qiita(keywords=["Python"], per_page=3, page=2)

    # 両方のページで結果が取得できた場合、記事IDが異なることを確認
    if len(articles_page1) > 0 and len(articles_page2) > 0:
        page1_ids = {article.id for article in articles_page1}
        page2_ids = {article.id for article in articles_page2}

        # 異なるページなので、記事が重複していないことを確認
        assert len(page1_ids & page2_ids) == 0, "Different pages should return different articles"


if __name__ == "__main__":
    print("統合テストを実行します...")
    print("注意: 実際のQiita APIを呼び出すため、レート制限に注意してください\n")

    asyncio.run(test_search_single_keyword())
    print("✓ 単一キーワードでの検索テスト完了\n")

    asyncio.run(test_search_multiple_keywords())
    print("✓ 複数キーワードでの検索テスト完了\n")

    asyncio.run(test_search_with_pagination())
    print("✓ ページネーションのテスト完了\n")

    print("すべての統合テストが完了しました")

File: examples/chapter02/mcp-server_qiita/typescript/src/integration-test.ts
import { searchQiita, QiitaArticle } from './qiita.js';

const sleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

const validateArticle = (article: QiitaArticle): boolean => {
  const requiredFields = [
    'id',
    'title',
    'url',
    'body',
    'created_at',
    'updated_at',
    'likes_count',
    'stocks_count',
    'user',
    'tags'
  ];

  for (const field of requiredFields) {
    if (!(field in article)) {
      console.error(`❌ 必須フィールド "${field}" が存在しません`);

      return false;
    }
  }

  if (typeof article.id !== 'string') {
    console.error('❌ id が文字列ではありません');

    return false;
  }

  if (typeof article.title !== 'string') {
    console.error('❌ title が文字列ではありません');

    return false;
  }

  if (typeof article.likes_count !== 'number') {
    console.error('❌ likes_count が数値ではありません');

    return false;
  }

  if (!article.user || typeof article.user.id !== 'string') {
    console.error('❌ user.id が不正です');

    return false;
  }

  if (!Array.isArray(article.tags)) {
    console.error('❌ tags が配列ではありません');

    return false;
  }

  return true;
};

const runIntegrationTest = async (): Promise<void> => {
  console.log('=== Qiita API 結合テスト開始 ===\n');

  let successCount = 0;
  let failureCount = 0;

  // テストケース1: 単一キーワード検索
  console.log('テストケース1: 単一キーワード検索');
  console.log('キーワード: ["TypeScript"]');

  try {
    const articles1 = await searchQiita(['TypeScript'], 5, 1);

    console.log(`✅ APIリクエスト成功: ${articles1.length}件の記事を取得`);

    if (!Array.isArray(articles1)) {
      console.error('❌ レスポンスが配列ではありません');
      failureCount++;
    } else if (articles1.length === 0) {
      console.log('⚠️  検索結果が0件でした');
    } else {
      console.log('\n記事情報:');

      for (const article of articles1.slice(0, 2)) {
        console.log(`  - タイトル: ${article.title}`);
        console.log(`    URL: ${article.url}`);
        console.log(`    いいね: ${article.likes_count}, ストック: ${article.stocks_count}`);
        console.log(`    タグ: ${article.tags.map((t) => t.name).join(', ')}`);
        console.log('');

        if (!validateArticle(article)) {
          failureCount++;

          break;
        }
      }

      successCount++;
    }
  } catch (error) {
    console.error('❌ エラーが発生しました:', error);
    failureCount++;
  }

  console.log('\n--- 3秒待機中... ---\n');
  await sleep(3000);

  // テストケース2: 複数キーワード検索
  console.log('テストケース2: 複数キーワード検索');
  console.log('キーワード: ["MCP", "API"]');

  try {
    const articles2 = await searchQiita(['MCP', 'API'], 3, 1);

    console.log(`✅ APIリクエスト成功: ${articles2.length}件の記事を取得`);

    if (!Array.isArray(articles2)) {
      console.error('❌ レスポンスが配列ではありません');
      failureCount++;
    } else if (articles2.length === 0) {
      console.log('⚠️  検索結果が0件でした');
    } else {
      console.log('\n記事情報:');

      for (const article of articles2.slice(0, 2)) {
        console.log(`  - タイトル: ${article.title}`);
        console.log(`    投稿者: ${article.user.name}`);
        console.log('');

        if (!validateArticle(article)) {
          failureCount++;

          break;
        }
      }

      successCount++;
    }
  } catch (error) {
    console.error('❌ エラーが発生しました:', error);
    failureCount++;
  }

  // 結果サマリー
  console.log('\n=== テスト結果 ===');
  console.log(`成功: ${successCount} / 2`);
  console.log(`失敗: ${failureCount} / 2`);

  if (failureCount === 0) {
    console.log('\n✅ 全てのテストが成功しました！');
  } else {
    console.log('\n❌ 一部のテストが失敗しました');
    process.exit(1);
  }
};

runIntegrationTest().catch((error) => {
  console.error('テスト実行中にエラーが発生しました:', error);
  process.exit(1);
});
